File: android/src/main/java/com/hlabexmaples/easyapi/ui/UsersFragmentWithViewModel.java
package com.hlabexmaples.easyapi.ui;

import android.arch.lifecycle.ViewModelProviders;
import android.support.design.widget.TabLayout;
import android.view.View;
import com.hlabexmaples.easyapi.R;
import com.hlabexmaples.easyapi.databinding.FragmentUsersBinding;
import com.hlabexmaples.easyapi.ui.base.BaseFragment;
import com.hlabexmaples.easyapi.ui.viewmodel.UserViewModel;

/**
 * Created by H.T. on 22/02/18.
 */

public class UsersFragmentWithViewModel extends BaseFragment<FragmentUsersBinding>
    implements TabLayout.OnTabSelectedListener {

  private UserListAdapter adapter;
  private UserViewModel viewModel;

  @Override
  protected int defineLayoutResource() {
    return R.layout.fragment_users;
  }

  @Override
  protected void initializeComponent(View view) {

    viewModel = ViewModelProviders.of(this).get(UserViewModel.class);

    adapter = new UserListAdapter();
    binding.rvUsers.setAdapter(adapter);

    binding.tab.addOnTabSelectedListener(this);

    viewModel.getUsersLiveData().observe(this, users -> {
      if (users != null && users.size() > 0) {
        hideLoading();
        adapter.setUserList(users);
      } else {
        showLoading();
      }
      binding.executePendingBindings();
    });
    observeLoadingState(viewModel.getLoadingStateLiveData());

    fetchUsers();
  }

  @Override
  protected void initToolbar() {
    assert getActivity() != null;
    binding.toolbar.setTitle(getString(R.string.title_users));
    binding.toolbar.setNavigationIcon(R.drawable.ic_back);
    binding.toolbar.setNavigationOnClickListener(view -> goBack());
  }

  @Override
  public void showLoading() {
    binding.setIsLoading(true);
  }

  @Override
  public void hideLoading() {
    binding.setIsLoading(false);
  }

  @Override
  public void onTabSelected(TabLayout.Tab tab) {
    int tag = tab.getPosition();
    switch (tag) {
      case 0:
        fetchUsers();
        break;
      case 1:
        fetchUsersRx();
        break;
    }
  }

  @Override
  public void onTabUnselected(TabLayout.Tab tab) {

  }

  @Override
  public void onTabReselected(TabLayout.Tab tab) {

  }

  private void fetchUsers() {
    viewModel.resetData();
    viewModel.fetchUsers();
  }

  private void fetchUsersRx() {
    viewModel.resetData();
    viewModel.fetchUsersRx();
  }

  /*Example*/
  private void fetchNextUsers() {
    viewModel.setNextPage();
    viewModel.fetchUsersRx();
  }

  @Override
  public void onClick(View view) {

  }

  @Override
  public void onDestroy() {
    super.onDestroy();
    viewModel.onDestroy();
  }
}

File: kotlin/src/main/java/com/hlabexamples/easyapi/kotlin/ui/UsersFragmentWithViewModel.kt
package com.hlabexamples.easyapi.kotlin.ui

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.support.design.widget.TabLayout
import android.view.View
import com.hlabexamples.easyapi.kotlin.data.models.User
import com.hlabexamples.easyapi.kotlin.ui.base.BaseFragment
import com.hlabexamples.easyapi.kotlin.ui.viewmodel.UserViewModel
import com.hlabexmaples.easyapi.R
import com.hlabexmaples.easyapi.databinding.FragmentUsersBinding

/**
 * Created by H.T. on 22/02/18.
 */

class UsersFragmentWithViewModel : BaseFragment<FragmentUsersBinding>(), TabLayout.OnTabSelectedListener {

    private lateinit var adapter: UserListAdapter
    private lateinit var viewModel: UserViewModel

    override fun defineLayoutResource(): Int {
        return R.layout.fragment_users
    }

    override fun initializeComponent(view: View) {

        viewModel = ViewModelProviders.of(this).get(UserViewModel::class.java)

        adapter = UserListAdapter()
        binding.rvUsers.adapter = adapter

        binding.tab.addOnTabSelectedListener(this)

        viewModel.usersLiveData.observe(this, Observer<List<User>> { users ->
            if (users != null && users.isNotEmpty()) {
                hideLoading()
                adapter.setUserList(users)
            } else {
                showLoading()
            }
            binding.executePendingBindings()
        })

        fetchUsers()
    }

    override fun initToolbar() {
        assert(activity != null)
        binding.toolbar.title = getString(R.string.title_users)
        binding.toolbar.setNavigationIcon(R.drawable.ic_back)
        binding.toolbar.setNavigationOnClickListener {
            goBack()
        }
    }

    override fun showLoading() {
        binding.isLoading = true
    }

    override fun hideLoading() {
        binding.isLoading = false
    }

    override fun onTabSelected(tab: TabLayout.Tab) {
        val tag = tab.position
        when (tag) {
            0 -> fetchUsers()
            1 -> fetchUsersRx()
        }
    }

    override fun onTabUnselected(tab: TabLayout.Tab) {

    }

    override fun onTabReselected(tab: TabLayout.Tab) {

    }

    private fun fetchUsers() {
        viewModel.resetData()
        activity?.let { viewModel.fetchUsers(it, this) }
    }

    private fun fetchUsersRx() {
        viewModel.resetData()
        activity?.let { viewModel.fetchUsersRx(it, this) }
    }

    /*Example*/
    private fun fetchNextUsers() {
        activity?.let {
            viewModel.setNextPage()
            viewModel.fetchUsersRx(it, this)
        }
    }

    override fun onClick(view: View) {

    }

    override fun onDestroy() {
        super.onDestroy()
        viewModel.onDestroy()
    }

}

